### PR TITLE
mgr/smb: add support for new ceph vfs module

### PR DESCRIPTION
**Pull request by phlogistonjohn**
Add support for using the new ceph vfs module in Samba (aka ceph_new) to the mgr module, and even default to using it for the samba-vfs provider. Add two new "sub-providers" one for forcing the old vfs module "samba-vfs/classic" and one, "samba-vfs/new", for forcing the new module regardless of what the default for "samba-vfs" is.

<s>TODO - Doesn't currently work correctly when using the new module and the "ceph:filesystem" config parameter. This needs to be resolved prior to merging.</s>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by anoopcs9** [comment]
> TODO - Doesn't currently work correctly when using the new module and the "ceph:filesystem" config parameter. This needs to be resolved prior to merging.

[!3748](https://gitlab.com/samba-team/samba/-/merge_requests/3748) created to address the issue.

File: src/pybind/mgr/smb/handler.py [hunk 1]
```diff
@@ -1039,11 +1043,16 @@ def _generate_share(
...
+    }[share.checked_cephfs.provider]
     cfg = {
         # smb.conf options
         'options': {
             'path': path,
-            "vfs objects": "acl_xattr ceph",
+            "vfs objects": f"acl_xattr {ceph_vfs}",
             'acl_xattr:security_acl_name': 'user.NTACL',
             'ceph:config_file': '/etc/ceph/ceph.conf',
             'ceph:filesystem': share.cephfs.volume,
```
Comment (anoopcs9) [inline src/pybind/mgr/smb/handler.py:1058]: With [!3751](https://gitlab.com/samba-team/samba/-/merge_requests/3751) merged you may now change the fixed `ceph` prefix.

**Review by anoopcs9 — CHANGES_REQUESTED** [review]
> > TODO - Doesn't currently work correctly when using the new module and the "ceph:filesystem" config parameter. This needs to be resolved prior to merging.
> 
> [!3748](https://gitlab.com/samba-team/samba/-/merge_requests/3748) created to address the issue.

Fix merged upstream.

File: src/pybind/mgr/smb/handler.py (hunk 1 above)
Comment (phlogistonjohn) [reply src/pybind/mgr/smb/handler.py:1058]: thanks for the heads up

**Comment by anoopcs9** [comment]
@phlogistonjohn Are we good to move ahead with an updated version of this change?

**Comment by phlogistonjohn** [comment]
I'm going to retest this once more manually and if all goes well I'll mark as reviewable.

**Comment by phlogistonjohn** [comment]
jenkins test api

**Review by anoopcs9 — APPROVED** [review]
lgtm.

**Comment by phlogistonjohn** [comment]
@adk3798 please include in a near-future teuth run. thanks!

**Review by adk3798 — APPROVED** [review]

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-09-10_02:59:23-orch:cephadm-wip-adk-testing-2024-09-09-1904-distro-default-smithi/

Failures:
- stray daemon warning in smb test, known issue
- `Test failure: test_nfs_export_creation_without_cmount_path (tasks.cephfs.test_nfs.TestNFS)` caused by a PR in the run
- 2x agent on staggered upgrade failure, known issue
- deploy raw test timed out, known issue
- `test_rgw_multisite` test failed to generate token on realm bootstrap, somewhat new but not the first time I've seen it and it's unrelated to PRs in this run
- 4x mds_upgrade_sequence failures, known issue
- rgw-ingress tests timed out, known issue

Nothing to block merging PRs in the run other than the NFS failure related one

**Pull request merged**